### PR TITLE
Extended Cache Selection Features

### PR DESCRIPTION
**Pull request by ajohnstonTE**
#### Multi-Method Selections

This forms the basis for a lot of the added features. Methods can now be selected on multiple times at once, ultimately forming a match condition akin to  `WHERE a AND b AND ...` in SQL. The "intersection" only utilizes the method value cache if _all_ of the methods are indexed.

A new, slightly more flexible/extensible syntax has also been added for performing selections. As a result of this change, code that previously would have looked like this (out of necessity):

```java
store().list(Foo.class, "getBarId", 5)
  .stream()
  .filter(foo -> foo.getDogId() == 6)
  .findFirst()
  .orElse(null);
```

can now become

```java
store().get(Foo.class, "getBarId", 5, "getDogId", 6);
```

or somewhat less error-prone as

```java
store().select(Foo.class)
  .where(MethodValue.of(Foo::getBarId, "getBarId").is(5))
  .where(MethodValue.of(Foo::getDogId, "getDogId").is(6));
  .get();
```

which avoids the var-args entirely. On that note...

#### Static-Type Checking in Selections

When using the new syntax, you are required to provide a function as the first parameter to each "where" method call. These are intended to be a method reference, specifically a method reference for the methods being selected on. Of course, due to limitations in Java, the method name string must still be provided. But it does allow for a) IDE's to suggest methods when typing out the method reference and b) static type analysis on the return type of the method to ensure the values being passed in are correct. This is particularly useful below (and not _quite_ as useful above, for reasons I'll get into).

#### Multi-Value Method Selections

Multiple values per methods can now be selected on, ultimately forming a match condition akin to  `WHERE a IN (...)` in SQL.

As a result of this change, code that previously would have looked like this (out of necessity):

```java
Stream.of(5, 8, 10)
  .flatMap(barId -> store()
      .list(Foo.class, "getBarId", barId)
      .stream()
      .filter(foo -> foo.getDogId() == 6))
  .distinct() // Assuming the equals method compares getId()
  .collect(Collectors.toList());
```

or, had support for selecting on multiple methods been added, like this

```java
Stream.of(5, 8, 10)
  .map(barId -> store().list(Foo.class, "getBarId", barId, "getDogId", 6))
  .flatMap(Collection::stream)
  .distinct() // Assuming the equals method compares getId()
  .collect(Collectors.toList());
```

can now be written in the normal syntax as

```java
store().get(Foo.class, "getBarId", new WhereInSet(Arrays.asList(5, 8, 10)), "getDogId", 6);
```

or more concisely/strongly-typed in the new syntax as

```java
store().select(Foo.class)
  .where(MethodValue.of(Foo::getBarId, "getBarId").in(Arrays.asList(5L, 8L, 10L)))
  .where(MethodValue.of(Foo::getDogId, "getDogId").is(6));
  .get();
```

which is a lot more clean (IMO). 

Note that I wrote it differently in the strongly-typed syntax, with explicit `long` markers for the where-in values. Supposing that Foo#getBarId actually returns a `long`, the way it is written in the normal syntax is actually completely wrong, and will never match anything. Java will auto-box the values in `Arrays.asList(5, 8, 10)` as Integers, rather than `Longs`, which will not equal the `Long` values stored in the `MethodValueCache`.

#### Multi-Entity Type Selections

Multiple entity types can also be selected on at the same time, which can be helpful should multiple separate entity classes have one or more interfaces/superclasses in common and a selection based on those requirements are needed. This can be done in one of two ways, each with their own benefits/drawbacks/use cases. However, neither offer the `get(long id)` method offered in the single entity type selector. It simply wouldn't make sense, as there is no way to know which type to select by id from. 

Anyways, suppose we had the following class definitions:

```java
interface Person extends Identifiable 
{ 
  long getId();
  String getName();
  int getAge();
}
interface Rich
{
  long getSalary();
}
abstract class Doctor implements Person, Rich
{
  abstract long getNumberOfLabCoats();
}
abstract class Lawyer implements Person, Rich
{
  abstract long getNumberOfSuits();
}
```

Here is the first syntax, in which each selected type is explicitly supplied:

```java
List<? extends Person> people = store().select(Arrays.asList(Doctor.class, Lawyer.class))
  .where(MethodValue.of(Person::getAge, "getAge").is(40))
  .where(MethodValue.of(Rich::getSalary, "getSalary").is(1_000_000_000L))
  .list()
```

In this syntax, a union of all common interfaces/superclasses in the list of entity classes is formed by the generic, allowing methods from both `Rich` and `Person` to be referenced despite being completely separate.

Here is the other syntax:

```java
List<Person> people = store().selectAnySubclass(Person.class)
  .where(MethodValue.of(Person::getAge, "getAge").is(40))
  // .where(MethodValue.of(Rich::getSalary, "getSalary").is(1_000_000_000L))
  .list()
```

The upside is that this syntax automatically finds any registered subclass of `Person` in the entity store and selects on it. Unfortunately, it also means that you cannot filter on `Rich` methods, but it still makes sense since that wasn't a requirement in the selector.

### Other Notes
- The code in MethodValueCache has been notably rewritten/expanded on (mostly expanded on), so tests were added for that class to make sure it was all still working correctly. 
  - The tests are admittedly a little hacky, as it creates a semi-fake application and cache, then provides the only required cache methods. But it did help find several bugs. All but the multi-method value tests were written and first tested on the old MethodValueCache code.
- I think the cache should absolutely log a warning every time a selection is made on a non-indexed method/intersection. I don't see *any* reason to do use the `.list(type, method, value)` entity store method instead of `.list(type).stream().filter(...)...` if you don't plan on indexing it aside from, maybe, to be concise.
- I recommend switching to JUnit 5, which I've already done in my other branches, for future tests. It's better for testing complex cases, or at least better for organizing them. I'll likely open up a PR at some point in the future for it.
- I recommend using [Testcontainer](https://www.testcontainers.org/)'s [database containers](https://www.testcontainers.org/modules/databases/) for simulating a full cache-to-db environment using dockerized SQL containers. 
  - I have this set up in another branch as well, with MySQL as the target container. My own usage of it is still pretty experimental and rough, but I used it to do some minor testing of the entity store methods I added here. I'll likely open up a PR at some point in the future for it.

### CODE REVIEW

**Comment by ajohnstonTE** [comment]
Renamed it to "Extended Cache Selection Features" because I ended up adding a few more features than the one I had originally planned. Also cleaned up the description to better explain each addition.

File: gemini/src/main/java/com/techempower/cache/EntitySelector.java [hunk 1]
```diff
@@ -26,14 +26,23 @@
...
    */
   public T get(long id)
   {
     return this.store.get(this.type, id);
   }
 
+  /**
+   * Returns the entity of the selected class with the given id. Ignores any
+   * specified method-value pairs.
+   */
```
Comment (msmith-techempower) [inline gemini/src/main/java/com/techempower/cache/EntitySelector.java]: Very minor nit - these comments are backwards - this one takes in a list of ids and calls for `the given id` while the one above takes a single id but calls for `the given ids`.

Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/cache/EntitySelector.java]: Whoops. Fixed.

**Comment by msmith-techempower** [comment]
@ajohnstonTE Can you merge master into this branch and push again so that the Travis tests will run?

**Comment by ajohnstonTE** [comment]
@msmith-techempower I can, but I'd rather not suggest that it's ready to merge. This PR is basically in WIP state because it's a super-massive memory hog. I need to revisit it at some point to try some other approaches that I came up with, but it hasn't been a priority recently. Especially since the cache isn't super useful to me atm.

Would you like me to close the PR and re-open it later if/when it's ready?

**Comment by ajohnstonTE** [comment]
Closed until the memory usage is fixed.

**Pull request closed without merging**